### DISCUSSION TITLE
`{ var ref  : T }` vs new `ref type`

### DISCUSSION THREAD

**Discussion by crusso** [caffeinelabs/motoko-core#59] [Ideas]
The impure collections wrap the pure collections in a single field object with a mutable field. That actually has more overhead than one would like in the current compiler (a boxed record with a named, boxed field). Perhaps it would be worth introducing a new stable ref type that just requires one indirection (not two) an always has constant time access rather than requiring a field lookup.

```
e  :  T
------------
ref e :  ref T

e : ref T
------------
! e : T

e1 : ref T   e2 : T
------------
e1 := e2 : ()
```

Unfortunately, doing this means adding a new representation to motoko values (i,e, a new tag similar to MutBox unless we just use MutBox for both fields and ref types).

**Comment by rvanasa** [comment]
Would it be feasible to optimize this as a special case (record with one `var` field) as an alternative to new language syntax?
